Added examples for 3 Authz request options (#218)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

: OPTIONAL. Defined in [@!RFC6749]. The Wallet MAY allow Verifiers to request presentation of Verifiable Credentials by utilizing a pre-defined scope value. See (#request_scope) for more details.

`response_mode`:
: OPTIONAL. Defined in [@!OAuth.Responses]. This parameter is used (through the new Response Mode `direct_post`) to ask the Wallet to send the response to the Verifier via an HTTPS connection (see (#response_mode_post) for more details). It is also used to request signing and encrypting (see (#jarm) for more details). If the parameter is not present, the default value is `fragment`.

The Verifier MAY send an Authorization Request using either of these 3 options:

> 1. Passing as URL with encoded parameters
> 2. Passing a request object as value
> 3. Passing a request object by reference

Reviewer: The formatting isn't working here, the list is coming out all on the same line:

![Screenshot 2024-07-23 at 21 37 49](https://github.com/user-attachments/assets/fcca1576-0512-43a6-9928-82340b5daedc)

I need it needs blank lines either side:

```suggestion

1. Passing as URL with encoded parameters
2. Passing a request object as value
3. Passing a request object by reference

```

Author: Hmm, the i use this VS code extension 
Link: https://marketplace.visualstudio.com/items?itemName=shd101wyy.markdown-preview-enhanced

And the earlier instance rendered fine.
<img width="706" alt="Screenshot 2024-07-23 at 2 04 00 PM" src="https://github.com/user-attachments/assets/c23d854a-16e1-4929-b487-d240040f4a32">

I can update it as per your suggestion.

For future reference can you share the editor/extension you use to render the `md` file?

Reviewer: The screenshot was of the official html rendering, which is generated using mmark - I downloaded it from the GitHub action tab, https://github.com/openid/OpenID4VP/pull/218/checks - by clicking on 'artifacts' and then 'output' here:

![Screenshot 2024-07-23 at 22 11 23](https://github.com/user-attachments/assets/be9c748e-ae8b-4922-956b-e6104634aee7)

Author: Thanks! will add this to the PR summary

> As defined in the JWT-Secured Authorization Request (JAR) [@RFC9101].

Reviewer 2: ```suggestion
2 and 3 are defined in the JWT-Secured Authorization Request (JAR) [@RFC9101].
```

it's only 2 and 3 that are defined in JAR, right?

Author: good catch! you are right. `presentation_definition` isn't defined in RFC9101

: OPTIONAL. Defined in [@!RFC6749]. The Wallet MAY allow Verifiers to request presentation of Verifiable Credentials by utilizing a pre-defined scope value. See (#request_scope) for more details.

`response_mode`:
: OPTIONAL. Defined in [@!OAuth.Responses]. This parameter is used (through the new Response Mode `direct_post`) to ask the Wallet to send the response to the Verifier via an HTTPS connection (see (#response_mode_post) for more details). It is also used to request signing and encrypting (see (#jarm) for more details). If the parameter is not present, the default value is `fragment`.

The Verifier MAY send an Authorization Request using either of these 3 options:

1. Passing as URL with encoded parameters
2. Passing a request object as value
3. Passing a request object by reference

2 and 3 are defined in the JWT-Secured Authorization Request (JAR) [@RFC9101].

The following is a non-normative example of Authorization Request with URL encoded parameters:
```
GET /authorize?
  response_type=vp_token
  &client_id=https%3A%2F%2Fclient.example.org%2Fcb
  &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
  &presentation_definition=...

>   &scope=openid

Reviewer: I'm not convinced it's a good idea to add scope=openid here, it makes it not a pure VP request and I think we want to show a basic VP request here.

Author: Should i remove it from all 3 examples then?

Reviewer: I think so, yes, please - it's not there in any of the examples currently in the spec.

Author: Done!

  &nonce=n-0S6_WzA2Mj HTTP/1.1
```

The following is a non-normative example of Authorization Request with request object as value:
```
GET /authorize?
  response_type=vp_token
  &client_id=https%3A%2F%2Fclient.example.org%2Fcb
  &request=eyJrdHkiOiJFQyIsImNydiI6IlAtMjU2IiwieCI6Ik1LQkNUTkljS1VTRGlpMTF5U3MzNTI2aURaOEFpVG83VHU2S1BBcXY3R
  DQiLCJ5IjoiNEV0bDZTUlcyWWlMVXJONXZmdlZIdWhwN3g4UHhsdG1XV2xiYk00SUZ5TSIsInVzZSI6ImVuYyIsImtpZCI6IjEiLCJhbGc

>   iOiJQUzI1NiJ9.eyJpc3MiOiJzNkJoZFJrcXQzIiwiYXVkIjoiaHR0cHM6Ly9zZWxmLWlzc3VlZC5tZS92MiIsInJlc3BvbnNlX3R5cGU

Reviewer 3: the header of this thing decodes to the following, which isn't ok 

{
  "kty": "EC",
  "crv": "P-256",
  "x": "MKBCTNIcKUSDii11ySs3526iDZ8AiTo7Tu6KPAqv7D4",
  "y": "4Etl6SRW2YiLUrN5vfvVHuhp7x8PxltmWWlbbM4IFyM",
  "use": "enc",
  "kid": "1",
  "alg": "PS256"
}

Author: I have removed the key and shortened the jwt to be eyJrd... as no key is specified.

Let me know if this looks good.

    "input_descriptors": [
      {
        "id": "id_credential",
        "format": {
          "jwt_vc_json": {
            "proof_type": [
              "JsonWebSignature2020"
            ]
          }
        },
        "constraints": {
          "fields": [
            {
              "path": [
                "$.vc.type"
              ],
              "filter": {
                "type": "array",
                "contains": {
                  "const": "IDCredential"
                }
              }
            }
          ]
        }
      }
    ]
  },
  "nonce": "n-0S6_WzA2Mj"
}
```
The example request object above is signed with this key:
```
{
  "kty": "EC",
  "crv": "P-256",
  "x": "MKBCTNIcKUSDii11ySs3526iDZ8AiTo7Tu6KPAqv7D4",
  "y": "4Etl6SRW2YiLUrN5vfvVHuhp7x8PxltmWWlbbM4IFyM",
  "use": "enc",
  "kid": "1",
  "alg": "PS256"
}
```

The following is a non-normative example of Authorization Request with request object as reference:
```
GET /authorize?
  &client_id=https%3A%2F%2Fclient.example.org%2Fcb
  &client_id_scheme=x509_san_dns
  &request_uri=https%3A%2F%2Fclient.example.org%2Frequest%2Fvapof4ql2i7m41m68uep
  &request_uri_method=post HTTP/1.1
  &scope=openid
```

> And, later the wallet might send the following non-normative example request to the `request_uri`:

Reviewer 4: ```suggestion
Later, the wallet might send the following non-normative example request to the `request_uri`:
```

  pbiFta_6CWrFxtC1QoCIrhIzbJgnF4GRboy0SfYRXD8Vt8qbPzLIeUvE-LyObGN5g
```
Where the contents of `request` consist of base64url-encoding and signing (in the example with RS256 algo)
this json:
```
{
  "iss": "s6BhdRkqt3",
  "aud": "https://self-issued.me/v2",
  "response_type": "vp_token",
  "client_id": "s6BhdRkqt3",
  "redirect_uri": "https//client.example.org/cb",
  "scope": "openid",
  "presentation_definition": {
    "id": "example_jwt_vc",
    "input_descriptors": [
      {
        "id": "id_credential",
        "format": {
          "jwt_vc_json": {
            "proof_type": [
              "JsonWebSignature2020"
            ]
          }
        },
        "constraints": {
          "fields": [
            {
              "path": [
                "$.vc.type"
              ],
              "filter": {
                "type": "array",
                "contains": {
                  "const": "IDCredential"
                }
              }
            }
          ]
        }
      }
    ]
  },
  "nonce": "n-0S6_WzA2Mj"
}
```
The example request object above is signed with this key:
```
{
  "kty":"RSA",

>   "kid":"k2bdc",

Reviewer 3: this key sure does bring back some memories, thanks for that, but this is not really consistent with how examples are provided in the rest of this document, is it?

Author: Updated to use existing key
```
{
  "kty": "EC",
  "crv": "P-256",
  "x": "MKBCTNIcKUSDii11ySs3526iDZ8AiTo7Tu6KPAqv7D4",
  "y": "4Etl6SRW2YiLUrN5vfvVHuhp7x8PxltmWWlbbM4IFyM",
  "use": "enc",
  "kid": "1",
  "alg": "PS256"
}
```

Reviewer 3: wait, what, the encryption key from one of the examples in the W3C Digital Credentials API section? that doesn't seem good. and where did you find the private key? so many questions...

Reviewer 3: wow the private key is apparently here https://datatracker.ietf.org/doc/html/rfc7517#appendix-A.2 which is fun

Reviewer 3: but maybe that's not even relevant

sorry

i am genuinely confused

Reviewer 3: and by confused https://github.com/openid/OpenID4VP/pull/218#discussion_r1693733995 i mean

Author: Thanks for your review and I hope to learn!

I found the key in this repo https://github.com/openid/OpenID4VP/blob/main/examples/digital_credentials_api/signed_request_payload.json#L19 and the draft https://openid.github.io/OpenID4VP/openid-4-verifiable-presentations-wg-draft.html
you are right it is the same one as in that doc in section A1 https://datatracker.ietf.org/doc/html/rfc7517#appendix-A.1 and A.2 section has the private key.

Lmk, if I should use a different key as its private key is readily available? Or do I need to create a new key pair and mention them in the doc? or something else?

Happy to get feedback and incorporate changes.

Reviewer 3: The idea of providing the key in a spec to allow for verification of examples is nice. But I don't think that's something that's been done much or at all in the 4VC family of documents. So doing it here is kind of out of place. And there are already plenty of opportunities for that in PAR, JAR, JWT, etc so it's arguably not necessary or even particularly useful to do here.  

It's pretty well accepted that using the same key pair for both singing and encryption is a bad and potentially dangerous thing. So we should avoid doing anything that looks like it's condoning or encouraging the practice.

Author: Thanks for the guidance! That makes sense!
I have removed the key and shortened the jwt to be `eyJrd...` as no key is specified.

Let me know if this looks good.

  &nonce=n-0S6_WzA2Mj HTTP/1.1
```

The following is a non-normative example of Authorization Request with request object as value:
```
GET /authorize?
    response_type=vp_token
    &client_id=https%3A%2F%2Fclient.example.org%2Fcb

>     &request=eyJhbGciOiJSUzI1NiIsImtpZCI6ImsyYmRjIn0.eyJpc3MiOiJzNkJoZFJrcXQzIiwiYXVkIjoiaHR0cHM6Ly9zZWxmLWlzc3VlZC5tZS92MiIsInJlc3BvbnNlX3R5cGUiOiJ2cF90b2tlbiIsImNsaWVudF9pZCI6InM2QmhkUmtxdDMiLCJyZWRpcmVjdF91cmkiOiJodHRwcy8vY2xpZW50LmV4YW1wbGUub3JnL2NiIiwic2NvcGUiOiJvcGVuaWQiLCJwcmVzZW50YXRpb25fZGVmaW5pdGlvbiI6eyJpZCI6ImV4YW1wbGVfand0X3ZjIiwiaW5wdXRfZGVzY3JpcHRvcnMiOlt7ImlkIjoiaWRfY3JlZGVudGlhbCIsImZvcm1hdCI6eyJqd3RfdmNfanNvbiI6eyJwcm9vZl90eXBlIjpbIkpzb25XZWJTaWduYXR1cmUyMDIwIl19fSwiY29uc3RyYWludHMiOnsiZmllbGRzIjpbeyJwYXRoIjpbIiQudmMudHlwZSJdLCJmaWx0ZXIiOnsidHlwZSI6ImFycmF5IiwiY29udGFpbnMiOnsiY29uc3QiOiJJRENyZWRlbnRpYWwifX19XX19XX0sIm5vbmNlIjoibi0wUzZfV3pBMk1qIn0.m1KYXT4r5Pj7Jh8gUlngpvjUoEZnOcbL8lXuDxbLbQyhLG9mro9EpT3I4YOm25tnxFx2mrKb42dFPxXOwLO5NbBmoYmY8wxH85H98RtpPGaTwCD6duxbp4eZm6LO-fdxpSvjzCAKoU3kzloh1Q1ijkP_ZQgwmS-uQba9qpu61hMjF6KIIjz59sJHpZOI0cEz4LqrLwXXIht0hyZXdE_vJIBEwfafoLZgggF9adNGUQnE3dXLi_is6XrDk7BS0VZds6IaNTVlIlS5NfHg6Lp1IpbiFta_6CWrFxtC1QoCIrhIzbJgnF4GRboy0SfYRXD8Vt8qbPzLIeUvE-LyObGN5g

Author: good catch!

```
Where the contents of `request` consist of base64url-encoding and signing (in the example with RS256 algo)
this json:
```
{
  "iss": "s6BhdRkqt3",
  "aud": "https://self-issued.me/v2",
  "response_type": "vp_token",
  "client_id": "s6BhdRkqt3",
  "redirect_uri": "https//client.example.org/cb",

>   "scope": "openid",

Reviewer: I'd have probably preferred if all three examples showed the same request so that we were just highlighting the differences of the 3 formats (e.g. this request has `scope=openid` and the above one doesn't) but I don't feel strongly about it.

Author: Good point! Agreed!

  &nonce=n-0S6_WzA2Mj HTTP/1.1
```

The following is a non-normative example of Authorization Request with request object as value:
```
GET /authorize?

>   response_type=vp_token

Reviewer: response_type isn't legal here, when `request=` is used response_type can only be inside the request object, it should be removed:
```suggestion
```

Author: Thanks! good catch!

```
GET /authorize?
  response_type=vp_token%20id_token

>   &scope=openid

Reviewer: This `scope=openid` should actually stay, as this section is showing an example of a SIOP request which does require `openid`. (openid should still be omitted in the new examples you're adding)

Author: Thanks! i wasn't sure about this earlier.
I've made the change.
